### ISSUE TITLE
libvarnish: VSS_resolver_socktype is not walking addrinfo correctly

### ISSUE THREAD

**Issue by asadsa92** [varnishcache/varnish-cache#3509]
commit da2409c3863d1da8f2dac0ace9ae05fa0e743d81 (master branch)

### One possible callgraph
```
func (== vtcp_open_callback)
VSS_resolver_socktype
VSS_resolver
VTCP_open
```

### The Problem
`VSS_resolver_socktype` checks the success of `func/vtcp_open_callback` call by simply:
```
	for (res = res0; res != NULL; res = res->ai_next) {
		vsa = VSA_Malloc(res->ai_addr, res->ai_addrlen);
		if (vsa != NULL) {
			ret = func(priv, vsa);
			free(vsa);
			if (ret)
				break;
		}
	}
```

But `vtcp_open_callback` `return (-1);` on error, `if (-1)` is true, so it bails out after just attempting the first ai_addr.

From `man getaddrinfo`, example code for client:
```
          for (rp = result; rp != NULL; rp = rp->ai_next) {
               sfd = socket(rp->ai_family, rp->ai_socktype,
                            rp->ai_protocol);
               if (sfd == -1)
                   continue;
               if (connect(sfd, rp->ai_addr, rp->ai_addrlen) != -1)
                   break;                  /* Success */               
               close(sfd);
           }
```
This is correct as it attempt to connect until it succeed.

**Comment by bsdphk** [comment]
Yes, that looks like a bug.

Not 100% sure if the correct fix is to make `VSS_resolver_socktype` continue on negative return values from the callback or for `vtcp_open_callback` to supress such values to zero.

**Comment by asadsa92** [comment]
Thanks, https://github.com/varnishcache/varnish-cache/commit/0d746632253d038c58110c9f2eeeef19b604556f fixes both #3509 and #3510.
Just a nitpick: `0` is a valid fd from `socket` when `stdin` has previously been closed.

**Comment by Dridi** [comment]
I think we `dup("/dev/null")` in place of `stdin`.
